src/core/RolesAuthority.sol

```diff
@@ -145,38 +150,72 @@ contract RolesAuthority is IAuthority, Initializable, UUPSUpgradeable {
         emit UserRoleUpdated(user, uint8(role), enabled);
     }
 
+    function setUserRole(address user, Role role, bool enabled) external virtual {
+        role == Role.System_FundAdmin ? _assertOwner() : _assertFundAdmin();
+
+        _setUserRole(user, role, enabled);
+
+        if (address(messenger) != address(0) && uint8(role) <= uint8(Role.Investor_Reserve5)) messenger.broadcast(msg.data);
+    }
+
+    function setUserRoleBatch(address[] memory users, Role[] memory roles, bool[] memory enabled, bool broadcast)
+        external
+        virtual
+    {
+        _assertFundAdmin();
+
+        uint256 length = users.length;
+        if (length == 0 || length != roles.length || length != enabled.length) revert InvalidArrayLength();
+
+        for (uint256 i; i < length;) {
+            if (uint8(roles[i]) > uint8(Role.Investor_Reserve5)) revert Unauthorized();
```

Comment:
@dsshap changed this from being `<=`, guess you meant to prevent batch from giving system roles